Documentation to Install nmos-cpp with Conan 2

johanjino opened the pull request.

johanjino committed:
Documentation to Install nmos-cpp with Conan 2

garethsb commented on Documents/Installation-with-Conan.md:
```diff
@@ -0,0 +1,57 @@
+## Guide to Install and Setup nmos-cpp using Conan Package Manager: 
+
+1. Install conan:
+`pip install --upgrade conan>=2.0.5`
```
- Using a code fence to get nicer formatting
- Indenting for better doc structure
- Changing from `>=` to `~=` because we wouldn't want an incompatible greater version, such as 3.0 might be

~~~suggestion
   ```sh
   pip install --upgrade conan~=2.0.5
   ```
~~~

garethsb commented on Documents/Installation-with-Conan.md:
```diff
@@ -0,0 +1,57 @@
+## Guide to Install and Setup nmos-cpp using Conan Package Manager: 
```
- no colon at end of heading
- shorter heading
- title case

```suggestion
## Install and Set Up nmos-cpp Using Conan

The following steps describe how to install and set up nmos-cpp using the Conan package manager.
For many platforms, a binary package is available from Conan Center Index so it isn't necessary to build nmos-cpp or any of its dependencies.
```

garethsb commented on Documents/Installation-with-Conan.md:
```diff
@@ -0,0 +1,57 @@
+1. Install conan:
+`pip install --upgrade conan>=2.0.5`
+
+    If conan.exe is not in PATH you will get warning: 
+	 
+
+	> WARNING: The script conan.exe is installed in 'C:\Users\%USER%\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\Scripts' which is not on PATH.
+	  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
+
+
+   Then the shown path needs to be added manually. After maunally adding the PATH, Restart the command line and run `conan --help` to ensure Conan is added to path successfully.
```
```suggestion
   If the python Scripts directory is not on the PATH you will get a warning like:
   > WARNING: The script conan.exe is installed in 'C:\Users\%USER%\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\Scripts' which is not on PATH.
   > Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.

   On Windows, you can use _System Properties \> Environment Variables..._ to permanently add the directory to the user PATH. Restart the Command Prompt and run `conan --help` to ensure Conan is found.
```

garethsb commented on Documents/Installation-with-Conan.md:
```diff
@@ -0,0 +1,57 @@
+
+   Then the shown path needs to be added manually. After maunally adding the PATH, Restart the command line and run `conan --help` to ensure Conan is added to path successfully.
+
+3. Detect Profile:
+	Run `conan profile detect`. \
+  This will create a conan profile with values based on your dev setup. Ideally the profile should be similar to these:
```
- isn't this 2 not 3?
- consistent capitalization, why "Profile" capitalized here, but Conan not in "Install conan"?

~~~suggestion
2. Detect profile:
   ```sh
   conan profile detect
   ```
   This will create a Conan profile with values based on your current platform, compiler, etc.
   Ideally the profile should be similar to these:
~~~

garethsb commented on Documents/Installation-with-Conan.md:
````diff
@@ -0,0 +1,57 @@
+	Run `conan profile detect`. \
+  This will create a conan profile with values based on your dev setup. Ideally the profile should be similar to these:
+	
+   For windows:
+	```{txt}
+	[settings]
+	arch=x86_64
+	build_type=Release
+	compiler=msvc
+	compiler.cppstd=14
+	compiler.runtime=dynamic
+	compiler.runtime_type=Release
+	compiler.version=193
+	os=Windows
+	```
+	For Linux:
+	```{txt}
+	[settings]
+	arch=x86_64
+	build_type=Release
+	compiler=gcc
+	compiler.cppstd=17
+	compiler.version=11
+	os=Linux
+	```
````
- spaces not tabs (also applies in other places)
- `ini` has nicer formatting for this sort of config file than `txt`
- removed the Windows `compiler.runtime_type` because I read somewhere (sorry, no reference) that it causes trouble?

~~~suggestion
   Windows
   ```ini
   [settings]
   arch=x86_64
   build_type=Release
   compiler=msvc
   compiler.cppstd=14
   compiler.runtime=dynamic
   compiler.version=193
   os=Windows
   ```
   Linux
   ```ini
   [settings]
   arch=x86_64
   build_type=Release
   compiler=gcc
   compiler.cppstd=17
   compiler.version=11
   os=Linux
   ```
~~~

garethsb replied:
- spaces not tabs (also applies in other places)
- `ini` has nicer formatting for this sort of config file than `txt`
- removed the Windows `compiler.runtime_type` because I read somewhere (sorry, no reference) that it causes trouble?

~~~suggestion
   Windows
   ```ini
   [settings]
   arch=x86_64
   build_type=Release
   compiler=msvc
   compiler.cppstd=14
   compiler.runtime=dynamic
   compiler.version=193
   os=Windows
   ```
   Linux
   ```ini
   [settings]
   arch=x86_64
   build_type=Release
   compiler=gcc
   compiler.cppstd=17
   compiler.version=11
   os=Linux
   ```
~~~

garethsb commented on Documents/Installation-with-Conan.md:
```diff
@@ -0,0 +1,57 @@
+	> Launch the registry using `./direct_deploy/nmos-cpp/bin/Release/nmos-cpp-registry.exe <your_config.json>` \
+	> Launch nodes using `./direct_deploy/nmos-cpp/bin/Release/nmos-cpp-node.exe <your_config.json>`.
+  
+  If your executables where added to PATH you could call then directly like:  `nmos-cpp-node <your_config.json>` and `nmos-cpp-registry <your_config.json>`
```
- fix typo "where" to "were"

```suggestion
4. Try running nmos-cpp-registry and/or nmos-cpp-node:
   If your executables were added to the PATH you could call then directly like:  `nmos-cpp-node <your_config.json>` and `nmos-cpp-registry <your_config.json>`
```

garethsb commented on Documents/Installation-with-Conan.md:
```diff
@@ -0,0 +1,57 @@
+	`conan install --tool-requires=nmos-cpp/cci.20240223`\
+  If on windows, also run `./conanbuild.bat` to ensure the executables for nmos-cpp-node and nmos-cpp-registry are added to path.
+
+   Alternatively if the nmos-cpp-node and nmos-cpp-registry has still not been added to path, you can add them manually (Executables likely located at C:\Users\%USERNAME%\.conan2\p\nmos-768c7905bf562\p\bin\Release) or work around this as shown below:
```
I think I don't want to get into the issue that _conanbuild.bat_ doesn't work, hopefully my PR on the CCI recipe is merged soon!

I'd be happy to keep the `direct_deploy` guidance as an alternative if tested on both Windows and Linux.

johanjino replied:
Tested and done: [commit](https://github.com/johanjino/nmos-cpp/pull/1/commits/45d43c12b36ff53973ed9e6998eb0da66ca49898)

garethsb commented on Documents/Installation-with-Conan.md:
```diff
@@ -0,0 +1,57 @@
+   	> - `--deployer=direct_deploy` copies the direct dependencies, in this case nmos-cpp-registry.exe and nmos-cpp-node.exe.
+   	>
+ 	> \
+	> Launch the registry using `./direct_deploy/nmos-cpp/bin/Release/nmos-cpp-registry.exe <your_config.json>` \
```
Oddly the documentation implies that the path includes the version as well as the dependency name, like **_direct_deploy/nmos-cpp/cci.20240223/bin/Release/nmos-cpp-registry_**.

Is that not the case?

https://docs.conan.io/2.0/reference/extensions/deployers.html#direct-deploy

johanjino replied:
![On my installation on windows](https://github.com/johanjino/nmos-cpp/assets/92362159/8b52a28b-b3dd-430b-a21d-877c9f8e5703)

This was how it was installed for me in Windows. Tried other packages are all of them did not have version. I later went to [conan installation script](https://github.com/conan-io/conan/blob/develop2/conan/internal/deploy.py#L118) and found this: 

```python
_deploy_single(dep, conanfile, output_folder, dep.ref.name)
```
Direct deploy only creates a folder with the ref.name, which here is nmos-cpp. On the other hand:

```python
folder_name = os.path.join("full_deploy", dep.context, dep.ref.name, str(dep.ref.version))
...
 _deploy_single(dep, conanfile, output_folder, folder_name)
```
Full deploy uses ref.version as well to generate folder name.


Seems like either a bug or an error in doc. Should I point this out as an GitHub issue?

garethsb replied:
Good research. ~Please open an [issue](https://github.com/conan-io/conan/issues) with the references to _deploy.py_ you included above, contrasted to the docs at https://github.com/conan-io/docs/blob/1757da950c00155b3410d9bcd1cb815a41f899e4/reference/extensions/deployers.rst?plain=1#L51.~

garethsb replied:
I think it's also instructive to read @memsharded guidance on virtual envs vs direct deploy - https://github.com/conan-io/conan/issues/15913#issuecomment-2014944542

memsharded replied:
I'd say some small difference among the deployers are expected:
- The ``direct_deploy`` is intended to deploy one package, without having to deploy its dependencies. Mostly for convenient extraction of some files inside a given package. So the files won't collide with other package files when deployed. The pkg name has been respected, but it could have been dropped too.
- The ``full_deploy`` deploys the full dependency graph. It is common that more than 1 different version of the same dependency occurs on a graph, then the version becomes more necessary to avoid collisions in the deployment.

garethsb replied:
@memsharded Makes sense, thanks! In which case, perhaps just the example in the documentation for the direct deployer is slightly wrong, since it includes a version directory.

memsharded replied:
Could be outdated or wrong, yep. If you want to do it, it would be great if you can submit a ticket or a PR to the docs about it, otherwise I can do it myself. Thanks for pointing this out!

garethsb replied:
Done: https://github.com/conan-io/docs/pull/3675

garethsb commented on Documents/Installation-with-Conan.md:
````diff
@@ -0,0 +1,57 @@
+	os=Linux
+	```
+
+4. Install nmos cpp:
+	`conan install --tool-requires=nmos-cpp/cci.20240223`\
+  If on windows, also run `./conanbuild.bat` to ensure the executables for nmos-cpp-node and nmos-cpp-registry are added to path.
````
- correct name ("nmos-cpp" not "nmos cpp")
- capitalize Windows
- fix incorrect slash for Command Prompt (unless we add instructions saying this is Powershell...?)
- add Linux equivalent

~~~suggestion
3. Install nmos-cpp:
   ```sh
   conan install --tool-requires=nmos-cpp/cci.20240223
   ```
   Make the `nmos-cpp-registry` and `nmos-cpp-node` executables available in the current session.
   On Windows, run `.\conanbuild.bat`.
   On Linux, run `./conanbuild.sh`.
~~~

johanjino committed:
Reformatting

- Added Linux equivalents
- Fixed typos

Co-authored-by: Gareth Sylvester-Bradley <31761158+garethsb@users.noreply.github.com>

johanjino committed:
Added and Verified Linux commands for direct deploy

- direct deploy Linux equivalent commands added
- Fixed formatting of %USERNAME%

garethsb commented on Documents/Installation-with-Conan.md:
```diff
@@ -0,0 +1,73 @@
+   On Linux, run `./conanbuild.sh`.
+
+  
+4. Try running nmos-cpp-registry and/or nmos-cpp-node:
+   If your executables were added to the PATH you could call then directly like:  `nmos-cpp-node <your_config.json>` and `nmos-cpp-registry <your_config.json>`
```
~~~suggestion
4. Try starting nmos-cpp-registry and/or nmos-cpp-node:
   ```sh
   nmos-cpp-registry
   nmos-cpp-node
   ```
   For more information about running these applications and the JSON configuration file that can be passed on the command-line, see the [Tutorial](Tutorial.md).
~~~

garethsb committed:
Add link to Tutorial

garethsb committed:
Omit workaround for bindirs issue that will be fixed soon

garethsb committed:
Link to Installation-with-Conan.md

garethsb committed:
Mention installing Python 3... could link to Dependencies.md#conan as well?

garethsb reviewed (approved):
LGTM - please make the PR on upstream sony/nmos-cpp
